drivers/pinctrl/pinctrl_xmc4xxx.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2022 Schlumberger
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/xmc4xxx-pinctrl.h>
+
+#include <devicetree.h>
+#include <xmc_gpio.h>
+
+#define REG_OR_NULL(node) COND_CODE_1(DT_NODE_EXISTS(node), (DT_REG_ADDR(node)), (NULL))
+
+static void * const gpio_ports[] = {
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio0)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio1)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio2)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio3)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio4)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio5)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio6)),
+};
+
+static int pinctrl_configure_pin(const pinctrl_soc_pin_t *pinctrl)
+{
+	int port_id, pin, alt_fun, hwctrl;
+	uint32_t pinmux = pinctrl[0];
+	XMC_GPIO_CONFIG_t pin_cfg = {0};
+
+	port_id = XMC4XXX_PINMUX_GET_PORT(pinmux);
+	if (port_id >= ARRAY_SIZE(gpio_ports) || !gpio_ports[port_id]) {
+		return -EINVAL;
+	}
+
+	pin = XMC4XXX_PINMUX_GET_PIN(pinmux);
+
+	if (XMC4XXX_PINMUX_GET_PULL_DOWN(pinmux)) {
+		pin_cfg.mode = XMC_GPIO_MODE_INPUT_PULL_DOWN;
+	}
+
+	if (XMC4XXX_PINMUX_GET_PULL_UP(pinmux)) {
+		pin_cfg.mode = XMC_GPIO_MODE_INPUT_PULL_UP;
+	}
+
+	if (XMC4XXX_PINMUX_GET_INV_INPUT(pinmux)) {
+		pin_cfg.mode |= 0x4;
+	}
+
+	if (XMC4XXX_PINMUX_GET_OPEN_DRAIN(pinmux)) {
+		pin_cfg.mode = XMC_GPIO_MODE_OUTPUT_OPEN_DRAIN;
+	}
+
+	if (XMC4XXX_PINMUX_GET_PUSH_PULL(pinmux)) {
+		pin_cfg.mode = XMC_GPIO_MODE_OUTPUT_PUSH_PULL;
+	}
+
+	alt_fun = XMC4XXX_PINMUX_GET_ALT(pinmux);
+	pin_cfg.mode |= alt_fun << PORT0_IOCR0_PC0_Pos;
+
+	/* only has effect if mode is push_pull */
+	if (XMC4XXX_PINMUX_GET_OUT_HIGH(pinmux)) {
+		pin_cfg.output_level = XMC_GPIO_OUTPUT_LEVEL_HIGH;
+	}
+
+	/* only has effect if mode is push_pull */
+	if (XMC4XXX_PINMUX_GET_OUT_LOW(pinmux)) {
+		pin_cfg.output_level = XMC_GPIO_OUTPUT_LEVEL_LOW;
+	}
+
+	/* only has effect if mode is push_pull */
+	pin_cfg.output_strength = XMC4XXX_PINMUX_GET_DRIVE(pinmux);
+
+	XMC_GPIO_Init(gpio_ports[port_id], pin, &pin_cfg);
+
+	hwctrl = XMC4XXX_PINMUX_GET_HWCTRL(pinmux);
+	if (hwctrl) {
+		XMC_GPIO_SetHardwareControl(gpio_ports[port_id], pin, hwctrl);
+	}
+
+	return 0;
+}
+
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt, uintptr_t reg)
+{
+	ARG_UNUSED(reg);
+	int ret = 0;
+
+	for (int i = 0; i < pin_cnt; i++) {
+		ret = pinctrl_configure_pin(pins++);
+		if (ret < 0) {
+			break;
+		}
+	}
+
+	return ret;
```

Comment:
```suggestion
	for (uint8_t i = 0U; i < pin_cnt; i++) {
		int ret = pinctrl_configure_pin(*pins++);
		if (ret < 0) {
			break;
		}
	}

	return 0;
```

Reply:
will update. Why do you prefer to return 0, and not forward the error?

Reply:
updated example, missed one return